src/php/Lang/Set.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+namespace Phel\Lang;
+
+use ArrayAccess;
+use Countable;
+use InvalidArgumentException;
+use Iterator;
+use Phel\Printer;
+
+class Set extends Phel implements Countable, Iterator, ICons, IPush, IConcat
+{
+
+    /**
+     * @var mixed[]
+     */
+    protected $data = [];
+
+    /**
+     * Constructor
+     *
+     * @param mixed[] $data A list of all values
+     */
+    public function __construct(array $data)
+    {
+        $this->setData($data);
+    }
+
+    /**
+     * Create a new set.
+     *
+     * @param mixed[] $values A list of all values
+     *
+     * @return Set
+     */
+    public static function create(...$values): Set
+    {
+        return new Set($values);
+    }
+
+    public function isTruthy(): bool
+    {
+        return true;
+    }
+
+    public function hash(): string
+    {
+        return spl_object_hash($this);
+    }
+
+    public function count(): int
+    {
+        return count($this->data);
+    }
+
+    public function current()
+    {
+        return current($this->data);
+    }
+
+    public function next()
+    {
+        next($this->data);
+    }
+
+    public function key()
+    {
+        return key($this->data);
+    }
+
+    public function valid()
+    {
+        return key($this->data) !== null;
+    }
+
+    public function rewind()
+    {
+        reset($this->data);
+    }
+
+    public function cons($x): ICons
+    {
+        array_unshift($this->data, $x);
+        return $this;
+    }
+
+    public function push($x): IPush
+    {
+        $hash = $this->offsetHash($x);
+        $this->data[$hash] = $x; // Don't need to check if $x is already there, just override.
+
+        return $this;
+    }
+
+    public function concat($xs): IConcat
+    {
+        foreach ($xs as $x) {
+            $this->push($x);
+        }
+
+        return $this;
+    }
+
+    public function union(Set $set): IConcat
+    {
+        return $this->concat($set->toPhpArray());
+    }
+
+    public function intersection(Set $set): IConcat
+    {
+        return $this->applySet($set, 'array_intersect');
```

Comment:
I think, we must call `array_intersect_key` here.

Why not just:
```
$this->data = array_intersect_key($this->data, $set->toPhpArray());
```
This is much simpler to read. Also you don't need the to wrap the method into a string.